misc/vtk/src/processors/vtktovolume.cpp

```diff
@@ -0,0 +1,406 @@
+        dataArrays_.addArray(arrayName, numComps, array->GetVoidPointer(0));
+        if (keepSettings) {
+            BoolProperty *prop = dynamic_cast<BoolProperty *>(dataArrays_[dataArrays_.size() - 1]);
+            if (prop && formerChannels.find(prop->getDisplayName()) != formerChannels.end())
+                prop->set(true);
+        }
+    }
+    if (!keepSettings && numTotalComps <= 4)
+        for (auto *prop : dataArrays_.getProperties())
+            if (BoolProperty *boolProp = dynamic_cast<BoolProperty *>(prop)) boolProp->set(true);
+
+    dataArrays_.selectedFormat = dataFormats_.get();
+
+    if (keepSettings) convertData();
+
+    reactToChange_ = true;
+}
+
+void VTKtoVolume::updateAvailableArrays() {
+    NetworkLock lock;
+    reactToChange_ = false;
+
+    int numTotalComps = 0;
+    auto properties = dataArrays_.getProperties();
+    for (size_t p = 0; p < properties.size(); ++p) {
+        if (dynamic_cast<BoolProperty *>(properties[p])->get())
+            numTotalComps += dataArrays_.numArrayComponents[p];
+    }
+
+    for (size_t p = 0; p < properties.size(); ++p) {
+        if (dynamic_cast<BoolProperty *>(properties[p])->get()) continue;
+        bool tooMuch = numTotalComps + dataArrays_.numArrayComponents[p] > 4;
+        properties[p]->setReadOnly(tooMuch);
+    }
+
+    reactToChange_ = true;
+}
+
+void VTKtoVolume::convertData() {
+    outport_.setData(nullptr);
+    if (!inport_.hasData()) return;
+
+    auto &vtkData = *inport_.getData();
+
+    // Get dimensions. If none are given, it's not a rectilinear grid.
+    std::optional<size3_t> dimensionsOpt = vtkData.getDimensions();
+    if (!dimensionsOpt) {
+        LogWarn("Input vtk data set is not rectilinear.");
+        return;
+    }
+    size3_t dimensions = *dimensionsOpt;
+    if (useCellData_.get())
+        for (size_t d = 0; d < 3; ++d) dimensions[d]--;
+
+    struct RawArrayData {
+        unsigned char *ptr;
+        size_t byteSize;
+    };
+    std::vector<RawArrayData> arrayDataPointers;
+
+    // How many components in total?
+    const DataFormatBase *format = DataFormatBase::get(dataFormats_.getSelectedValue());
+    size_t sizeElement = format->getSize();
+    size_t numTotalComps = 0;
+    for (size_t p = 0; p < dataArrays_.size(); ++p) {
+        BoolProperty *prop = dynamic_cast<BoolProperty *>(dataArrays_[p]);
+        if (prop && prop->get()) {
+            arrayDataPointers.push_back(
+                {dataArrays_.ptrArrayData[p], dataArrays_.numArrayComponents[p] * sizeElement});
+            numTotalComps += dataArrays_.numArrayComponents[p];
+        }
+    }
+    if (!numTotalComps) {
+        LogWarn("No components selected.");
+        return;
+    }
+
+    const DataFormatBase *multichannelFormat =
+        DataFormatBase::get(format->getNumericType(), numTotalComps, format->getSize() * 8);
+
+    // Use point bounds
+    Matrix<4, float> worldMatrix(0);
+    double bounds[6];
+    vtkData->GetBounds(bounds);
```

Comment:
Does this return the min/max of voxel centers or voxel "boxes"? I guess it is the whole voxels and not the centers, if thats true, the code is correct.

Reply:
It is the min/max of the point positions, so it should do what is expected.

Reply:
I've checked with a script I used to convert some vtk data, it retuns the bounds voxel centered.

Reply:
Furthermore, this will only work when the data is aligned with the xyz coordinate system.

Reply:
The following is the code I use in my script:
```c++

vtkStructuredGrid* grid = ...;

vtkPoints* points = grid->GetPoints();

util::IndexMapper3D index(dims);

dvec3 p0,px,py,pz;

points->GetPoint(0, glm::value_ptr(p0));
points->GetPoint(index(1, 0, 0), glm::value_ptr(px));
points->GetPoint(index(0, 1, 0), glm::value_ptr(py));
points->GetPoint(index(0, 0, 1), glm::value_ptr(pz));

const auto dx = px - p0; // world offset vector between two neighboring voxels (on the grids x-axis)  
const auto dy = py - p0; // world offset vector between two neighboring voxels (on the grids y-axis)
const auto dz = pz - p0; // world offset vector between two neighboring voxels (on the grids z-axis)

dmat3 basis = {
    dx * dims.x,
    dy * dims.y,
    dz * dims.z
};

dvec3 offset = p0 - (dx + dy + dz) * 0.5; // p0 is the center of the voxel, offset needs another half a voxel 

volume->setBasis(basis);
volume->setOffset(offset);
```

Reply:
That seems like the better solution for uniform grids. It makes things worse for rectilinear grids, but at least the rotation will be correct. I'll update it to your variant.

Reply:
Well volume does not support rectilinear data.. so that not really relevant